Sacc 2

joezuntz opened the pull request:
Hi all, esp. @beckermr @elikrause @tmcclintock.

This is a draft pull request to get some comments before moving things around and doing a proper one.

This is my attempt, in the sacc2 directory, to implement a new sacc format.  This is considerably less highly structured than sacc.  The idea is to make something that is extensible into whatever other types of data that we want.

There are two new example notebooks for creating this format.

A few notes
- each data point is represented by a small struct containing the mean value, tracers (e.g. bins), the type of data (e.g. shear_ee) and a list of tags.  This tag list contains things like binning information or windows.
- I/O is just done using the python pickle module, since this doesn't need to be high performance and it's simple
- there are a bunch of tools and helper methods for selecting or cutting out subsets of data
- there are several helper methods specifically for 2pt data since that's our first use case.  We can add more for other data types without modifying the format.
- tracers are looked up by a name, e.g. 'source_1'
- the window code is as-yet not developed but the idea is to supply a window as a tag.  Pickle knows to only save windows once if they're used multiple times.

Thoughts?

LATER UPDATES
- Switched to a FITS output
- Added windows properly

joezuntz commented:
@slosar meant to tag you as well.

beckermr commented:
One quick comment. I'm against adopting pickle as the default format because changes in the code can render old pickle files unreadable or they can be read incorrectly. I'll take a closer look next week.

joezuntz commented:
@beckermr - that's a good point, thanks.  I'll have a think about other format possibilities.

beckermr commented:
My vote is actually this https://github.com/dask/fastparquet. This format will compress repeated entries in columns, like tags etc really well. It is very readable from python. The overall dev has moved here: https://github.com/apache/arrow

Then again, we could just use fits. :P

damonge commented:
I'd prefer a more standard format (i.e. FITS or HDF) so people aren't put off using it because they need to learn something new

beckermr commented:
Arguable arrow/parquet is used by more people than FITS, but yeah I totally get that. I don't care much. CSV is fine by me. Just not pickle for this purpose.

joezuntz commented:
Not sure how well FITS and HDF will cope with potentially somewhat heterogenous data.  It might be a bit awkward.

Of course we could enforce more homogeneity.  I agree you want something standard-ish within astro/cosmo.

slosar commented:
I think you are reinventing a lot, but at the end of the day, it is your judgement and your time...
I do agree that pickle is not a good choice. We could try pickling by enforcing some very early version and keep it compatible, but it just sounds too fragile. I'm also strongly against CSV, it is pain and you *will* get people writing covariance matrix elements with insufficient precision.
fits/hdf5/parquet I'm agnostic about, though I also vote for FITS.
It was actually on my todo list to raise this issue next week when I'm back. I think we should make a list of all people that use sacc in some sort, which at the moment is lss for the HSC project, joe for 3x2pt and perhaps some people in SO via damonge and then just make an executive decision on whether to fix old sacc or list of things that need to be implemented in sacc2. Given that I'm using this least, so it should be really be decided between mat, joe and david...

slosar commented:
Is YAML a good thing? I'm always worried about cov matrices in ASCII, the most constraining directions are the smallest corrections in cov matrix, so very small changes in numbers can make big changes in C^{-1}.

joezuntz commented:
It's bad if people try to write them by hand, yes, but if it's done using the tools here then no precision could be lost - if you just do `print(x)` in python then maximum precision will always be written out.

That's why I'm keen for people to use a library for generating these files, rather than do things by hand.

damonge commented:
My concern as well (sorry, a lot of the conversation happened on slack).
Does the ease of coding for yaml compensate for the bad storage and I/O efficiency?
Or are there other benefits for users that I'm not seeing?

joezuntz commented:
Ease of coding is the main advantage, yes.  Also it's flexible.

joezuntz commented:
Let's have the conversation on slack, though, I find it much easier to follow.  we can put some conclusions here.

damonge commented on sacc2/sacc.py:
```diff
@@ -0,0 +1,1026 @@
+
+
+# add more as we develop them
+allowed_types = [
```
I don't see why we need to specify `galaxy`, `shear` or `ggl`. These are just different combinations of spin-0 and spin-2 two-point functions. E.g. why should we add another type when we want to incorporate CMB lensing x delta_gal correlations? That should be encompassed by what currently is `galaxy_density_cl` or `galaxy_density_w`, i.e. the correlation of two spin-0 fields.

joezuntz replied:
Yes, I'm not 100% sure about this.  We could store things just in terms of the two fields, as we did in the 2pt code.  But there are some combinations like magnification which can be hard to distinguish from others if you just specify the two tracers and their types.  So I'm really not sure.  Also if you switch to a different statistic like mass aperture then you need a new name.

damonge replied:
Can you point me to one example of a magnification data vector, just so I understand this better?
The way I thought about this, at least in the context of an N-point function, is that the tracers tell you the quantity that you're correlating (number counts, magnification, shear, whatever), while the type tells you the combination of spin components you're correlating and whether it's in real or Fourier space. I think this should be enough to specify everything, but I may be missing something.

Since a user can always define a new tracer with some degree of freedom, this is then easy to extend to any quantity we may care about in cosmology without having to add a bunch of new types every time we want to include a new observable (since in general we will want to include all its possible correlations with all existing observables).

joezuntz replied:
I was just meaning that if you do mag x mag then it's another spin-0 galaxy correlation, just like counts x counts.  But you're right we can include another piece of info about what it is.

What about things like COSEBIs that are combinations of other 2pt functions?

But as I say, I'm really not sure what the best way of doing this is!

slosar replied:
In original sacc we have an additional concept of a physical tracer, i.e.  a set of galaxies, from which you can derive shear or number counts, but in retrospect I agree that that was a bad idea (I had some vague notion about correlated systematics, but now I think there are probably better ways of doing that). So, once tracer = one field on the sky. But quantity is a separate thing, not only there are real and config, there are also various compensated statistc, etc as you say.

joezuntz replied:
So I've tried to summarize this in my head, and have come to these two options:

data_type is a string
- pro: less to pass around when creating (slightly simpler)
- pro: slightly faster searches (not a big difference)
- con: need to create new ones for new combinations, more verbose


data_type is a pair of strings
- pro: don't have to create a new name when a new type of data is added
- pro: potentially more explicit about what things are
- pro: perhaps easier to automate code that uses the two individually
- con: non-2pt data does not fit this model at all! e.g. cluster things.

To me the last of these is the most important - if we want this to work for non-2pt data then we need to keep the data type as a string, at the risk of increasing the number of combination unnecessarily for 2pt data.

joezuntz replied:
Have I missed something?  If not I'll add any that we can think of now and put it into a separate python file so it's easier to add things later.

beckermr replied:
It has to be a string - my intention is to use this format for everything

damonge replied:
Wait, but data_type in this case would be `number_counts` or `one_point` or something like that. It'd be associated to a single tracer, corresponding to the number of clusters in a given redshift-richness bin.

This is the way we thought about structuring things when we discussed this particular problem with Elisabeth and Hironao.

damonge replied:
This is also good in the sense that it allows you to include all two-point functions between any observable and a particular cluster bin (you just reference the corresponding cluster `tracer`).

beckermr replied:
Not following @damonge. Note that I want to use this format for MCMC chains too. I mean literally everything.

joezuntz replied:
The data_type string should let you know what calculation you need to do to predict the value.  The tracers and window (etc)  then contain metadata for that calculation.

This is *definitely* not designed for MCMC chains!  They are always homogenous, so much more suited to existing formats.

damonge replied:
The way we thought about this: the data_type tell you about the maths of the data point: whether it's an N-point function, what spin components, Fourier or real. The tracers that each data point is associated with then tell you everything you need to know about the physical quantities involved. You need both pieces of information in order to compute the prediction.

Structuring things this way is good, I think, because it compartimentalises things better, and it prevents you from defining N^D data types if you are considering N different tracers and up to D-point functions. You only need to define N tracers, and those can actually be fully defined by the user.

damonge replied:
Yes, please, let's not use this for MCMC, catalogs, images, TODs or anything else ;-)

joezuntz replied:
The maths of the data point isn't enough - it needs to tell us the physics as well.
For example, CMB shear fields and galaxy shear fields have the same mathematical structure, but we're feeding this data into firecrown, which needs to know whether to use a delta function n(z) or a distribution, which changes how you do the integral (or don't do it!) completely.  So the data_type needs to tell us what entire calculation you have to do, so that means it should be something like galaxy_shear_ee rather than just "2pt_spin2_ee" or something like that.

beckermr replied:
Designed vs used is a very different statement @joezuntz :p

damonge replied:
I don't think so. In this particular case the data_type would be `ClEE`, and the two tracers would be a shear tracer (identified by a tag in it that says it's shear, and containing an N(z)), and a CMB lensing tracer (identified by a tag in it that says it's CMB lensing, and not much else).

The tracers just need to have tags that identify them, and those tags don't have to be standardized (i.e. users could create new tracers if they need them). Whatever code reads the file just needs to be able to interpret these tags (firecrown in this case).

joezuntz replied:
I think we just disagree here - I'd like this to be more explicit and for someone looking at the data to be able to tell explicitly at a glance at the row what kind of data it is. And you'd prefer something defined more mathematically and with minimal repetition by encoding this information in the tracer.

Anyone else have any thoughts?

damonge replied:
yes, I think that sums it up ;-)

In my favour, I think this setup would also make your life easier in firecrown: for a given datapoint, the tracers tell you directly what pair of CCL objects you need to correlate. Otherwise you'll need to write a dictionary that maps datatypes to pairs of tracers.

damonge replied:
Since the initial discussion about cluster implementation (which seems to be the outlier here) was initiated with @elikrause , it might be worth pinging her.

slosar replied:
For what it is worth, I tend to agree with David here. At the end of the day, we'll have manually marginalise over all kinds of crap, so firecrown will likely do things differently for different galaxy shear, i.e. you don't want to have a special data type for red and blue galaxies, just because intrinsic alignments is different. So let this field just describe the quantity and the recipe for calculation comes from the particular tracer...

joezuntz replied:
@beckermr your thoughts here would be useful too.

joezuntz replied:
Okay, I think I'm just going to make a decision here and stick with what I've done so far, as ultimately it's not that important either way, and I appointed myself the sacc champion at Becker's demand.   Does anyone have any more comments?  I will document the code better and ask for a code review after that.

damonge replied:
I think this will make things difficult when it comes to bringing in more probes or when using this for other experiments. Then again, since I haven't done the work, I don't want to block any progress.

That being said, I will feel free to submit a PR with what I think is be the best way of doing this later on ;-)

joezuntz replied:
Opening an issue for this.

damonge commented on sacc2/sacc.py:
```diff
@@ -0,0 +1,1026 @@
+allowed_types = [
+    "shear_xi_plus",
+    "shear_xi_minus",
+    "shear_ee",
```
Also, we may want to store EB

joezuntz replied:
Will add!  I'm expecting that we'll gradually add more data types as time goes on; that's why the code doesn't cause an error if you use an unknown type, just a warning.

joezuntz replied:
I've added EB.  We'll need to make a fuller list of expected data types.

damonge commented on sacc2/sacc.py:
```diff
@@ -0,0 +1,1026 @@
+# add more as we develop them
+allowed_types = [
+    "shear_xi_plus",
+    "shear_xi_minus",
```
In all generality, these things have imaginary components caused by parity-breaking correlations (i.e. EB power spectra). You may want to include those, in the same way you're including `ggl_gamma_x`, which is the imaginary part of the ggl correlation function.

joezuntz replied:
added

damonge commented on sacc2/sacc.py:
```diff
@@ -0,0 +1,1026 @@
+        ----------
+        mask: array or list
+            Mask must be either a boolean array or a list/array of integer indices to remove.
+            if boolean then True means to cut data point and False means to keep it
```
This seems exactly the opposite of what one would expect (i.e. mask=1 usually means keep - because you're not masking it). You would also save a couple of `not` operations below.

joezuntz replied:
I think I made it this way round because the method was called "cut", so to me that meant the True=cut.  But yes, when you have a mask that normally means the other way around.  I'll make a new method called "mask" and rename the argument in this one, so that we have both options.

joezuntz replied:
I've added both variants now.

damonge commented on sacc2/sacc.py:
```diff
@@ -0,0 +1,1026 @@
+        if not len(mask)==len(self):
+            raise ValueError("Mask passed in is wrong size")
+
+        self.data = [d for i,d in enumerate(self.data) if not mask[i]]
```
I might worry that this kind of loops might make things slow (as opposed to a simple evaluation of an ndarray on `mask`). Has this been benchmarked?

damonge replied:
There are other instances of this (e.g. in `reorder`)

joezuntz replied:
Running reorder on 2000 data points takes 4ms.  It should scale as O(n log(n)).

joezuntz replied:
The sort, I mean.

damonge replied:
ok, no problem then!

damonge commented on sacc2/sacc.py:
```diff
@@ -0,0 +1,871 @@
+        else:
+            return angle, mu
+
+    def get_ell_cl(self, data_type, bin1, bin2, return_cov=False):
```
Same comment as above for the case of swapping bin1 and bin2

damonge replied:
Maybe also worth noting that these functions cannot be used to retrieve covariance matrix elements between two different power spectra. This might limit their usefulness.

joezuntz replied:
Yeah, I was really only thinking that return_cov would be useful for plotting.

damonge replied:
sure, and that's fine

damonge commented on sacc2/sacc.py:
```diff
@@ -0,0 +1,871 @@
+        data_type: str or array/list of str
+            Which type C_ell to add
+        bin1: str or array/list of str
+            The name(s) of the first tomographic bin
```
the mention of tomographic bins here seem too specific to me. We want this to be used for other things (cmb lensing, etc.). In order to keep the notation homogeneous with the rest of the code, I'd change the docstrings and rename bin1, bin2 to tracer1, tracer2. Same for `_add_2pt`, `add_theta_xi`, and the corresponding `get` functions.

joezuntz replied:
These are convenience functions for 3x2pt, where the tracer does correspond to a tomographic bin.  Do you think that the name change is confusing?

damonge replied:
OK, but the function would be exactly the same if you wanted to retrieve power spectra between CMB kappa and delta_g, right? Or between any other pair of maps. So it'd make sense to have one single function for this, rather than copies of the same function with different documentation, right?

I think this brings me back to the interpretation of data_type as something that is separate from the physical properties of the tracers you're correlating (the argument is the same there: you don't need to define different types for power spectra between different types of tracers)

joezuntz replied:
That's a good point - I'll switch to tracer1, tracer2 and put a note in the docs that tomographic bins are one example.

joezuntz commented on sacc2/sacc.py:
```diff
@@ -0,0 +1,871 @@
+        """
+        return copy.deepcopy(self)
+
+    def to_canonical_order(self):
```
I just thought it would be useful to have some kind of default ordering, to make it easier to compare data vectors.

joezuntz commented on sacc2/sacc.py:
```diff
@@ -0,0 +1,871 @@
+        # Since we don't want to re-order the file as a side effect
+        # we first make a copy of ourself and re-order that.
+        S = self.copy()
+        S.to_canonical_order()
```
It's not 100% necessary, but I thought it would be useful since then two files should look much more similar to each other if they have the same content.

damonge replied:
Right, I don't disagree. There are however applications where having a different ordering is more optimal (e.g. to fully account for the non-Gaussian likelihood of the C_ells on large scales a la https://arxiv.org/abs/0801.0554 it's better to have the bin indices be the fastest-growing ones, rather than ell). So could we make this reordering optional through a keyword argument that defaults to `True`?

joezuntz replied:
sounds good - will do.

joezuntz replied:
Actually I've just remembered the other reasoning here - we need to group together the data points which share the same data_type, since they will be saved in a single table together.  If we don't do this then the covariance will be in the wrong order.

You can imagine a user e.g. alternating adding EE and BB data points for the same ell, but when we save we need to re-group them.

damonge replied:
OK, I see. I didn't know these were being saved in different tables. I guess forget about this then.

joezuntz commented on sacc2/sacc.py:
```diff
@@ -0,0 +1,871 @@
+
+    def cut(self, cut):
+        """
+        Remove data points, getting rid of points where the mask is True or an index is
```
Yep, exactly - I was thinking of calls to `indices` followed by `cut` or mask.
Definitely a good idea to have something that does both together, since it's a common pattern - maybe rename `cut` to `cut_indices` and have a `cut` method that does both?

damonge replied:
Sounds good! Having specific "scale_cut" functions where you just pass an array of ell_min and ell_max for each C_ell (and thetas for correlation functions) would also be useful. And a similar one for "cut all points involving tracer X".

joezuntz replied:
I've done the first part of this, renaming the functions and adding.  I'd like to punt a little on the helper functions in the interests of getting through this, though they're definitely handy.

damonge replied:
ok, sounds good!

joezuntz commented on sacc2/sacc.py:
```diff
@@ -0,0 +1,871 @@
+        """
+        return self.get_tags([tag], data_type=data_type, tracers=tracers, **select)[0]
+
+    def get_data_points(self, data_type=None, tracers=None, **select):
```
I was thinking of cases where you have more rich tags than we have for 3x2pt, to avoid the need to do repeated searches in get_tag.  But that might be a bit farfetched.  Could maybe delete this or make it internal only?

damonge replied:
We can leave it there in case it's useful (and delete later if it turns out not to be). Just wanted to understand the logic.

joezuntz commented on sacc2/sacc.py:
```diff
@@ -0,0 +1,871 @@
+        indices = self.indices(data_type=data_type, tracers=tracers, **select)
+        return [self.data[i] for i in indices]
+
+    def get_mean(self, data_type=None, tracers=None, **select):
```
Yeah, you have very few guarantees about this, especially when you start thinking of cross-sample things (e.g. an EE power spectrum could be non-symmetric if it's between two samples).  One way to deal would be just having a convention.

damonge replied:
Another option is to just warn the user of this, so they make sure to pass all possible combinations of things they care about in different calls to this function...

joezuntz replied:
Added.  I forgot to mention that the get_tracer_combinations method is meant to help with this.

joezuntz commented:
> Allow them to have arbitrary metadata as a dictionary that gets stored in some HDU header.

Definitely agree on this - I will add.

> Allow them to optionally store a 2D array of size [ncol,nrow] with extra information. 

I'm not so sure about this - I was imagining that tracers with substantial new functionality would be new subclasses.  Do you think that's too complicated or hard for people?


> Also, until we have converged on the best use of the data_type - tracers combination, I'd rather suppress the warning about unknown datatypes.

Good idea, will do.

> Another minor thing. I get a warning when reading:

That's an issue with h5py itself - they are fixing it in the next release.

damonge commented:
> > Allow them to optionally store a 2D array of size [ncol,nrow] with extra information.
> 
> I'm not so sure about this - I was imagining that tracers with substantial new functionality would be new subclasses. Do you think that's too complicated or hard for people?
Mmh, I was thinking of this as a generic feature that all tracers should have in order to have a more flexible format: an additional, optional storage space (which would be empty by default) made up of a bunch of columns to pass additional, arbitrary information around. We had this in the original sacc through an `extra_columns` argument (which now could be a FITS Table HDU). Example: for the HSC project we have 5 different estimates of the N(z) from the data, but we don't need to generate 5 versions of the SACC file for this, we just store the fiducial N(z) as the tracer's N(z), and then all other versions of it as extra columns with names attached to them.
This would allow anyone to use the miscellaneous tracer as almost any new tracer they want without having to code it up, make a PR etc.

> > Another minor thing. I get a warning when reading:
> 
> That's an issue with h5py itself - they are fixing it in the next release.
Ah right, sorry.

slosar commented:
>  I don't like this being completely arbitrary
Are you referring to the extra table that David requested? I think the point is that you never know what you will need so at some level it is useful to have a general bag of stuff to carry around to help you with housekeeping and let the final user have freedom. If it turns out that the same specific functionality is req'd repeatedly, you make it part of the official API and standardize its use.

damonge commented:
Yes, that's what I meant

joezuntz commented:
Yes, that's what I meant as well.  I'd like to avoid everyone coming up with their own different and incompatible generic extensions to objects here if possible, and put them into the library directly sooner rather than later.  A general bag of metadata is already supported, but for larger bits of data that accompany it I'd like to be explicit.

joezuntz commented:
Happy to consult more widely, though.

joezuntz commented:
I think this is ready to be converted to a full pull request now, so I'm going to do that.

damonge commented:
> Yes, that's what I meant as well. I'd like to avoid everyone coming up with their own different and incompatible generic extensions to objects here if possible, and put them into the library directly sooner rather than later. A general bag of metadata is already supported, but for larger bits of data that accompany it I'd like to be explicit.

I don't see the problem with adding an extra miscellaneous bag of data. You never know which extra data you want to pass around (e.g. magnification bias parameter, multiplicative biases as a function of z, angle-dependent weight for star obscuration, depth/redshift-dependent completeness, who knows). Many of these things are specific enough that you don't want to go through the fuss of writing a PR and waiting for it to be merged into the main code (or even editing the code at all) before using them, because it's not necessarily something that needs to be widely supported. So if there's no extra bag to put random information, I worry that people will just write it into separate ASCII files, which then may lead to inconsistencies in the analysis codes (which is precisely the kind of thing we wanted to solve with SACC).

slosar commented:
I agree with the last comment. Not having a functionality just because you are afraid people might abuse it, it is not a good argument for not having it. You should give people hard time when they start abusing it, not before.

joezuntz commented:
Getting back to Sacc now after a break.

I've updated the more_nz to be called extra_columns instead in the NZTracer, to implement what people have suggested above about extra bags of data.  It's not completely general, but should certainly cover all the things people have suggested above, and I'm happy to revisit later if needed.

Need to decide if we are just calling this sacc with a bumped version number or sacc2 - the former is easier.  I'll ask this on the HSC channel as well.

beckermr commented on sacc2/sacc.py:
```diff
@@ -0,0 +1,940 @@
+        tracers: tuple of str
+            Strings corresponding to zero or more of the tracers in the data set
+            These should either be already set up using the add_tracer method,
+            or you could set tracers_laster=True if you want to add them later.
```
spelling

joezuntz replied:
fixed.

beckermr commented on sacc2/sacc.py:
```diff
@@ -0,0 +1,940 @@
+        if self.covariance is not None:
+            self.covariance = self.covariance.masked(indices)
+
+    def cut_indices(self, indices):
```
This seems confusing to me. What is wrong with asking people to write `~mask` in the method before this one?

beckermr commented on sacc2/sacc.py:
```diff
@@ -0,0 +1,940 @@
+            m[i] = True
+        return m
+
+    def mask_indices(self, indices):
```
So this function masks where the input mask is False? That def seems backwards. Maybe name it `select_indices` or something?

joezuntz replied:
Blame @slosar and @damonge ! How about `keep_indices` and `remove_indices`?

beckermr replied:
Yup. keep and remove are great.

beckermr replied:
TBH, we only need one.

joezuntz replied:
Changed to keep and remove.  It's no extra maintenance cost to support both since one just calls the other.

beckermr commented on sacc2/utils.py:
```diff
@@ -0,0 +1,159 @@
+    """
+    M = np.atleast_2d(M)
+
+    # Our strict mode will only work 
```
comment seems not done

joezuntz replied:
fixed.

joezuntz commented on setup.py:
```diff
@@ -10,3 +10,11 @@
       packages=['sacc'])
+
+
+setup(name="sacc2", 
```
Done

joezuntz commented on test/test_sacc2.py:
```diff
@@ -0,0 +1,134 @@
+import sacc2
```
Done throughout.

joezuntz commented on sacc2/covariance.py:
```diff
@@ -0,0 +1,484 @@
+
+from .utils import invert_spd_matrix
+
+class BaseCovariance:
```
I was aiming for python3 only.  That seems like it will be okay right now?  Happy to reconsider if neeed.

beckermr replied:
Seems fine to me!

joezuntz commented on sacc2/covariance.py:
```diff
@@ -0,0 +1,484 @@
+
+
+
+        """
```
Have dealt with this throughout.

joezuntz commented on sacc2/covariance.py:
```diff
@@ -0,0 +1,484 @@
+            If a list, the total length of all the arrays in it
+            should equal n.  If an array, it should be either 1D of
+            length n or 2D of shape (n x n).
+        n: int
```
That's currently used as a check that the covariance has the right size, but it probably doesn't make sense to do that in the constructor.  I'll move it somewhere more sensible.

joezuntz replied:
Done.

joezuntz commented on sacc2/data_types.py:
```diff
@@ -0,0 +1,251 @@
+        """Create a new data point.
+
+        Data points can be automatically created and added to a
+        Sacc object, so you don't normally nee to manually create them.
```
Fixed.

joezuntz commented on sacc2/data_types.py:
```diff
@@ -0,0 +1,251 @@
+        #     warnings.warn(f"Unknown data_type value {data_type}. If possible use a pre-defined type, or add to the list.")
+
+    def __repr__(self):
+        return f"<Data {self.data_type} {self.tracers} {self.value} {self.tags}>s"
```
Nice - done.  It won't work 100% (because it would be impractical to make window function objects eval in the same way, since they can be complicated) but it should mostly work.

joezuntz commented:
Okay, I've addressed all of the comments from @beckermr above, adding a test for more of the selection tools, changing the verbs to keep/remove and standardising the docstrings.

I've tagged the previous version with 0.1.0, and once we've merged we can tag the new one.

It looks like the changes to deleted files on master mean this can't be merged automatically yet - once people are happy with this I will look into that.

joezuntz commented:
Just realised I missed a few comments - looking at these now.

joezuntz commented on sacc2/covariance.py:
```diff
@@ -0,0 +1,484 @@
+        hdu.header['sacctype'] = 'cov'
+        hdu.header['saccclss'] = self.cov_type
+        hdu.header['size'] = self.size
+        hdu.header['blocks'] = len(self.blocks)
```
There is a test that the covariance is correctly recovered.

joezuntz commented on sacc2/covariance.py:
```diff
@@ -0,0 +1,484 @@
+
+        """
+        if np.any(np.diff(indices))<0:
+            raise ValueError("Indices passed to BlockDiagonalCovariance.get_block must be in ascending order")
```
No, it's unsafe, because you would normally use this to select a chunk of the mean as well as the matrix, so if you just reordered that matrix it would return inconsistent data.

So the caller needs to sort them (and in the places we call it already this is automatic).

joezuntz replied:
It could be rewritten to allow arbitrary orderings, but the code and the resulting matrix would be very messy.  It seemed better to disallow this.

joezuntz commented on sacc2/data_types.py:
```diff
@@ -0,0 +1,251 @@
+    "galaxy_shear_eb": ["ell"],
+    "galaxy_density_cl": ["ell"],
+    "ggl_E": ["ell"],
+    "ggl_B": ["ell"],
```
Do you mean the ones here: 
https://ccl.readthedocs.io/en/v1.0.0/api/pyccl.correlation.html#pyccl.correlation.correlation
?
We definitely need more than that since we want to measure data types that are always zero in the theory predictions.

Something more consistent would be better than what's there now, though.

joezuntz replied:
Resolving for now

beckermr replied:
agreed we need more - we should fix this sooner rather than later since it is a core part of the api and future changes will be painful.

beckermr commented on sacc/covariance.py:
```diff
@@ -0,0 +1,466 @@
+        return cls(C)
+
+
+    def masked(self, mask):
```
We should change this verb too.

joezuntz replied:
How about `new_cov = cov.keeping(indices)`?

beckermr replied:
or `keep_indices` like the other one?

joezuntz replied:
`keep_indices` is in-place (and so imperative), so it makes sense to use a different verb form (tense? mood?) for this, which returns a new instance.

joezuntz replied:
Maybe `keeping_indices`

beckermr replied:
Ahhhh ok. No reason to obsess over this `with_only_indices` would do the trick or something similar.

beckermr replied:
yup. `keeping_indices` is fine.

joezuntz replied:
Changed.  And test added.

beckermr commented on sacc/covariance.py:
```diff
@@ -0,0 +1,466 @@
+            s += sz
+        return scipy.linalg.block_diag(*sub_blocks)
+
+    def masked(self, mask):
```
Ditto here.

beckermr commented on sacc/covariance.py:
```diff
@@ -0,0 +1,466 @@
+        hdu.header['saccclss'] = self.cov_type
+        return hdu
+
+    def masked(self, mask):
```
Ditto here.

joezuntz commented:
Only thing left now is the type names, and we said we'd review this in the future anyway as we learn how people use the code.  The current choice is designed to be human readable more than anything else.  We'll definitely need to add to the list, whatever happens.

So are people happy if I resolve the last few conversations above and merge?

joezuntz commented:
I re-requested @beckermr to review, since I think that's how to make the red circle go away.

damonge commented on sacc/data_types.py:
```diff
@@ -0,0 +1,374 @@
+    "clusterGalaxy_densityShear_xi_t": ["theta"],
+    "clusterGalaxy_densityShear_xi_x": ["theta"],
+    "clusterGalaxy_densityShear_xi_rt": ["radius"],
+    "clusterGalaxy_densityShear_xi_rx": ["radius"],
```
So here's a list of missing tags (`<...>` means all possible cl/xi/cosebi):
```
galaxy_convergenceDensity_<...>
clusterGalaxy_density_<...>
cmbGalaxy_temperatureDensity_<...>
cmbGalaxy_polarizationDensity_<...>
galaxy_shearConvergence_<...>
clusterGalaxy_densityShear_cosebi<...>
cmbGalaxy_temperatureShear_<xi/cosebi...>
cmbGalaxy_polarizationShear_<...>
clusterGalaxy_densityConvergence_<...>
cmbGalaxy_temperatureConvergence_<...>
cmbGalaxy_polarizationConvergence_<...>
cluster_density_<...>
cmbCluster_temperatureDensity_<...>
cmbCluster_polarizationDensity_<...>
cmb_temperaturePolarization_<xi/cosebi>
cmb_polarization_<xi/cosebi>
```
Some of these have zero expectation value, but can be null tests (the same way that you have shear_eb, for example). Others are not mainstream (e.g. CMB correlation functions), but they can definitely be computed. Others are more usual than some of the ones you've added (e.g. I'd argue that `cmbGalaxy_temperatureDensity` is more likely to be interesting than `cmbGalaxy_temperatureShear`).

A few questions:
- do `galaxy_density_cosebi`s make sense (same for convergence, cluster_density, cmb_temperature etc.)?
- what do `xi_eplus` and `xi_bplus` mean? and what's the difference with `xi_plus` (same for `minus`)?
- should there be a `cmb_convergence` type for CMB lensing? That cross-correlation seems to me more important than the CMB temperature/polarization ones.
- why do you have  `galaxy_shearDensity_cl_e` but `clusterGalaxy_densityShear_cl_te`?
- what is `clusterGalaxy_densityShear_xi_rt`? what is the difference with `clusterGalaxy_densityShear_xi_t`?
- `cmb_temperaturePolarization_cl_t` is repeated twice
- You can have `cmbGalaxy` but not `galaxyCmb`  (and many similar others). Will this make it confusing?

I think there should also be some guidance for what the user should do if they want to include data points that are not supported here. E.g. for systematics checks we will definitely want to compute galaxy-star correlation functions, or galaxy-systematic correlation functions (where "systematic" is any diagnostic map, same for shear).

joezuntz replied:
Many thanks!  I've added the missing ones here.

> do galaxy_density_cosebis make sense (same for convergence, cluster_density, cmb_temperature etc.)?

- No!  Deleted

> what do xi_eplus and xi_bplus mean? and what's the difference with xi_plus (same for minus)?

- I was thinking of the E-only and B-only xi values.  But they're rarely useful - I've deleted.

> should there be a cmb_convergence type for CMB lensing? That cross-correlation seems to me more important than the CMB temperature/polarization ones.

- added, thanks!

> why do you have galaxy_shearDensity_cl_e but clusterGalaxy_densityShear_cl_te?

- changed to canonical order 

> what is clusterGalaxy_densityShear_xi_rt? what is the difference with clusterGalaxy_densityShear_xi_t?
- I was thinking of xi(r) vs xi(theta).  But not useful or clear I think - deleted and have put an example of doing this in a notebook

> cmb_temperaturePolarization_cl_t is repeated twice

- thanks, deleted

> You can have cmbGalaxy but not galaxyCmb (and many similar others). Will this make it confusing?

- I think better to pick one of the two?

> I think there should also be some guidance for what the user should do if they want to include data points that are not supported here

- good point - added to the docs.

damonge commented on sacc/data_types.py:
```diff
@@ -0,0 +1,404 @@
+    "galaxy_shear_cosebi_ee": [],
+    "galaxy_shear_xi_minus": ['theta'],
+    "galaxy_shear_xi_plus": ['theta'],
+}
```
Thanks a lot for the documentation for additional types!

About the official ones, sorry, more nags!
I think you're missing
```
cmbCluster_convergenceDensity_<cl/xi>
cmb_polarizationConvergence_<cl/xi>_<e/b/t/x>
cmb_temperatureConvergence_<cl/xi>
cmbGalaxy_polarizationConvergence_xi_<t/x>
```

And questions:
- why clusterGalaxy_densityShear_cl_tb but cmbCluster_polarizationDensity_cl_b? I think there are others like this. I think you're aiming for spin0-spin2 correlation functions to be xi_t and xi_x.
- Similar: what is "cmbGalaxy_temperatureShear_xi_tb" and "cmbGalaxy_temperatureShear_xi_te"? I guess they should be "cmbGalaxy_temperatureShear_xi_t/x"?
- In principle, spin2-spin2 things like "cmbGalaxy_polarizationShear_xi_minus" have imaginary components corresponding to the parity-violating terms (C_ell^EB). E.g. see Eq. 13 in https://arxiv.org/pdf/astro-ph/0303414.pdf
- You have "cmb_temperaturePolarization_cl_t" only, but you should have "cmb_temperaturePolarization_cl_e" and "cmb_temperaturePolarization_cl_b".
- I think "cmb_temperature_cl_tt" should just be "cmb_temperature_cl"
- You only have cosebis for shear. Should they be there for CMB polarization too?

damonge replied:
I'm fine with keeping the imaginary parts out if you want to, BTW, just thought it was worth bringing it up, since the EB/BE power spectra are there.

joezuntz replied:
> I think you're missing

Thanks, added them all!

> why clusterGalaxy_densityShear_cl_tb but cmbCluster_polarizationDensity_cl_b?

Fixed in that case, and the others

> what is "cmbGalaxy_temperatureShear_xi_tb" and "cmbGalaxy_temperatureShear_xi_te"? 

Thanks - fixed (remove b/e)

> In principle, spin2-spin2 things like "cmbGalaxy_polarizationShear_xi_minus" have imaginary components

Added imagPlus and imagMinus variants - might be useful.

> You have "cmb_temperaturePolarization_cl_t" only, but you should have "cmb_temperaturePolarization_cl_e" and "cmb_temperaturePolarization_cl_b".

Changed.

> I think "cmb_temperature_cl_tt" should just be "cmb_temperature_cl"

Agreed - changed.

>  You only have cosebis for shear. Should they be there for CMB polarization too?

Added!

damonge replied:
Thanks @joezuntz . Two tiny ones:

You have `cmbGalaxy_temperatureShear_xi_t` twice. I think the second one should be `cmbGalaxy_temperatureShear_xi_x`

Also, if `polarization` and `shear` have cosebis, then `polarizationShear` should probably have them too (although I don't care enough about cosebis to push this one).

joezuntz replied:
Thanks - fix pushed for both

damonge replied:
Thanks!

joezuntz commented:
Okay, manually merging via command line, following the instructions from github!
